File: scripts/generate_meta_assets.py
#!/usr/bin/env python3
"""
Generate Meta Assets (Favicon & OG Image) for Siquijor.xyz
==========================================================

Creates favicon and OpenGraph image from scraped Siquijor images.

Usage:
    python scripts/generate_meta_assets.py

Requirements:
    pip install Pillow
"""

import os
from pathlib import Path
from PIL import Image, ImageDraw, ImageFont, ImageFilter

# Paths
BASE_DIR = Path(__file__).parent.parent
PUBLIC_DIR = BASE_DIR / "public"
IMAGES_DIR = PUBLIC_DIR / "images"
SCRAPED_DIR = IMAGES_DIR / "scraped"

# Colors - Siquijor theme (teal/turquoise waters)
SIQUIJOR_TEAL = (0, 169, 157)  # #00A99D
SIQUIJOR_DARK = (26, 60, 64)   # #1A3C40
WHITE = (255, 255, 255)


def create_favicon():
    """Create a simple favicon with Siquijor's initial 'S'."""
    sizes = [16, 32, 48, 64, 128, 180, 192, 512]

    # Create the largest size first
    size = 512
    img = Image.new('RGBA', (size, size), (0, 0, 0, 0))
    draw = ImageDraw.Draw(img)

    # Draw circular background with gradient effect
    # Main circle
    padding = size // 10
    draw.ellipse([padding, padding, size - padding, size - padding],
                 fill=SIQUIJOR_TEAL)

    # Add a subtle inner glow
    inner_pad = padding + size // 20
    draw.ellipse([inner_pad, inner_pad, size - inner_pad, size - inner_pad],
                 fill=(0, 189, 177))  # Slightly lighter teal

    # Draw the 'S' letter
    try:
        # Try to use a system font
        font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", size // 2)
    except (OSError, IOError):
        try:
            font = ImageFont.truetype("/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf", size // 2)
        except (OSError, IOError):
            font = ImageFont.load_default()

    text = "S"
    # Get text bounding box
    bbox = draw.textbbox((0, 0), text, font=font)
    text_width = bbox[2] - bbox[0]
    text_height = bbox[3] - bbox[1]

    # Center the text
    x = (size - text_width) // 2
    y = (size - text_height) // 2 - size // 15

    draw.text((x, y), text, fill=WHITE, font=font)

    # Save different sizes
    for s in sizes:
        resized = img.resize((s, s), Image.Resampling.LANCZOS)

        if s == 16:
            resized.save(PUBLIC_DIR / "favicon-16x16.png")
        elif s == 32:
            resized.save(PUBLIC_DIR / "favicon-32x32.png")
        elif s == 180:
            resized.save(PUBLIC_DIR / "apple-touch-icon.png")
        elif s == 192:
            resized.save(PUBLIC_DIR / "android-chrome-192x192.png")
        elif s == 512:
            resized.save(PUBLIC_DIR / "android-chrome-512x512.png")

    # Create ICO file
    ico_sizes = [(16, 16), (32, 32), (48, 48), (64, 64)]
    ico_images = [img.resize(s, Image.Resampling.LANCZOS) for s in ico_sizes]
    ico_images[0].save(PUBLIC_DIR / "favicon.ico", format='ICO', sizes=ico_sizes)

    print(f"✓ Created favicons in {PUBLIC_DIR}")


def find_best_og_image():
    """Find the best landscape image for OG meta."""
    candidates = []

    # Look for beach or general images (usually best for OG)
    search_dirs = [
        SCRAPED_DIR / "paliton-beach",
        SCRAPED_DIR / "siquijor-general",
        SCRAPED_DIR / "cambugahay-falls"
    ]

    for search_dir in search_dirs:
        if search_dir.exists():
            for ext in ['*.jpg', '*.jpeg', '*.png', '*.webp']:
                for img_path in search_dir.glob(ext):
                    try:
                        with Image.open(img_path) as img:
                            w, h = img.size
                            # Prefer landscape images close to 1200x630 ratio
                            ratio = w / h
                            if 1.5 <= ratio <= 2.1 and w >= 800:
                                candidates.append((img_path, w, h, abs(ratio - 1.9)))
                    except Exception:
                        continue

    # Sort by ratio closeness to 1.9 (OG ideal)
    if candidates:
        candidates.sort(key=lambda x: x[3])
        return candidates[0][0]

    return None


def create_og_image():
    """Create OpenGraph image with overlay branding."""
    og_width, og_height = 1200, 630

    # Find a good base image
    base_path = find_best_og_image()

    if base_path:
        print(f"Using base image: {base_path.name}")
        img = Image.open(base_path)

        # Resize to cover OG dimensions
        img_ratio = img.width / img.height
        og_ratio = og_width / og_height

        if img_ratio > og_ratio:
            # Image is wider - scale to height
            new_height = og_height
            new_width = int(new_height * img_ratio)
        else:
            # Image is taller - scale to width
            new_width = og_width
            new_height = int(new_width / img_ratio)

        img = img.resize((new_width, new_height), Image.Resampling.LANCZOS)

        # Center crop
        left = (new_width - og_width) // 2
        top = (new_height - og_height) // 2
        img = img.crop((left, top, left + og_width, top + og_height))

        # Convert to RGB if necessary
        if img.mode != 'RGB':
            img = img.convert('RGB')
    else:
        # Create a gradient background if no image found
        print("No suitable image found, creating gradient background")
        img = Image.new('RGB', (og_width, og_height), SIQUIJOR_TEAL)

    # Add semi-transparent overlay at bottom
    overlay = Image.new('RGBA', (og_width, og_height), (0, 0, 0, 0))
    draw = ImageDraw.Draw(overlay)

    # Gradient overlay at bottom for text
    for y in range(og_height // 2, og_height):
        alpha = int(180 * (y - og_height // 2) / (og_height // 2))
        draw.line([(0, y), (og_width, y)], fill=(26, 60, 64, alpha))

    # Composite overlay
    img = Image.alpha_composite(img.convert('RGBA'), overlay)

    draw = ImageDraw.Draw(img)

    # Add text
    try:
        title_font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 64)
        subtitle_font = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 28)
    except (OSError, IOError):
        try:
            title_font = ImageFont.truetype("/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf", 64)
            subtitle_font = ImageFont.truetype("/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf", 28)
        except (OSError, IOError):
            title_font = ImageFont.load_default()
            subtitle_font = ImageFont.load_default()

    # Title
    title = "Siquijor.xyz"
    title_bbox = draw.textbbox((0, 0), title, font=title_font)
    title_x = 60
    title_y = og_height - 160

    # Shadow
    draw.text((title_x + 2, title_y + 2), title, fill=(0, 0, 0, 150), font=title_font)
    draw.text((title_x, title_y), title, fill=WHITE, font=title_font)

    # Subtitle
    subtitle = "Your Guide to the Mystical Island"
    draw.text((60, og_height - 85), subtitle, fill=(255, 255, 255, 230), font=subtitle_font)

    # Convert back to RGB and save
    img = img.convert('RGB')
    img.save(PUBLIC_DIR / "og-image.jpg", quality=90, optimize=True)
    img.save(IMAGES_DIR / "og-default.jpg", quality=90, optimize=True)

    print(f"✓ Created OG image: {PUBLIC_DIR / 'og-image.jpg'}")


def create_site_webmanifest():
    """Create web manifest for PWA support."""
    manifest = """{
  "name": "Siquijor.xyz",
  "short_name": "Siquijor",
  "icons": [
    {
      "src": "/android-chrome-192x192.png",
      "sizes": "192x192",
      "type": "image/png"
    },
    {
      "src": "/android-chrome-512x512.png",
      "sizes": "512x512",
      "type": "image/png"
    }
  ],
  "theme_color": "#00A99D",
  "background_color": "#ffffff",
  "display": "standalone"
}"""

    with open(PUBLIC_DIR / "site.webmanifest", "w") as f:
        f.write(manifest)

    print(f"✓ Created web manifest")


def main():
    print("\n" + "=" * 60)
    print("  GENERATING META ASSETS")
    print("=" * 60 + "\n")

    # Create directories if needed
    IMAGES_DIR.mkdir(parents=True, exist_ok=True)

    create_favicon()
    create_og_image()
    create_site_webmanifest()

    print("\n" + "=" * 60)
    print("  META ASSETS GENERATED SUCCESSFULLY")
    print("=" * 60 + "\n")


if __name__ == "__main__":
    main()

File: scripts/generate-meta-assets.js
#!/usr/bin/env node
/**
 * Generate Meta Assets (Favicon & OG Image) for Siquijor.xyz
 * Uses sharp.js for image processing
 *
 * Usage:
 *   npm install sharp
 *   node scripts/generate-meta-assets.js
 */

import sharp from 'sharp';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const BASE_DIR = path.resolve(__dirname, '..');
const PUBLIC_DIR = path.join(BASE_DIR, 'public');
const IMAGES_DIR = path.join(PUBLIC_DIR, 'images');
const SCRAPED_DIR = path.join(IMAGES_DIR, 'scraped');

// Brand colors matching tailwind.config.mjs
const JUNGLE_700 = '#1A4D2E';  // Primary deep jungle green
const JUNGLE_600 = '#2d6d5a';
const AMBER_400 = '#F2A922';   // Accent golden amber
const CREAM = '#FDFCF8';       // Background cream
const VOLCANIC_900 = '#18181b'; // Dark mode background

/**
 * Create favicon from scratch using SVG
 */
async function createFavicon() {
  console.log('Creating favicons...');

  // Create SVG for favicon - jungle green with amber accent leaf icon
  const svgFavicon = `
    <svg width="512" height="512" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
      <!-- Background circle with jungle green -->
      <circle cx="256" cy="256" r="240" fill="${JUNGLE_700}"/>
      <!-- Inner circle slightly lighter -->
      <circle cx="256" cy="256" r="210" fill="${JUNGLE_600}"/>
      <!-- Stylized leaf/palm icon -->
      <g transform="translate(256, 256) scale(3.5)" fill="none" stroke="${CREAM}" stroke-width="1.8" stroke-linecap="round" stroke-linejoin="round">
        <path d="M13 8c0-2.76-2.46-5-5.5-5S2 5.24 2 8h2l1-1 1 1h4"/>
        <path d="M13 7.14A5.82 5.82 0 0 1 16.5 6c3.04 0 5.5 2.24 5.5 5h-3l-1-1-1 1h-3"/>
        <path d="M5.89 9.71c-2.15 2.15-2.3 5.47-.35 7.43l4.24-4.25.7-.7.71-.71 2.12-2.12c-1.95-1.96-5.27-1.8-7.42.35"/>
        <path d="M11 15.5c.5 2.5-.17 4.5-1 6.5h4c2-5.5-.5-12-1-14"/>
      </g>
      <!-- Amber accent dot -->
      <circle cx="380" cy="140" r="35" fill="${AMBER_400}"/>
    </svg>
  `;

  const sizes = [
    { name: 'favicon-16x16.png', size: 16 },
    { name: 'favicon-32x32.png', size: 32 },
    { name: 'apple-touch-icon.png', size: 180 },
    { name: 'android-chrome-192x192.png', size: 192 },
    { name: 'android-chrome-512x512.png', size: 512 },
  ];

  // Create base image from SVG
  const baseBuffer = Buffer.from(svgFavicon);

  for (const { name, size } of sizes) {
    await sharp(baseBuffer)
      .resize(size, size)
      .png()
      .toFile(path.join(PUBLIC_DIR, name));
    console.log(`  ✓ Created ${name}`);
  }

  // Create ICO file (just use the 32x32 PNG as favicon.ico)
  await sharp(baseBuffer)
    .resize(32, 32)
    .png()
    .toFile(path.join(PUBLIC_DIR, 'favicon.ico'));
  console.log('  ✓ Created favicon.ico');
}

/**
 * Find the best landscape image for OG from scraped images
 */
async function findBestOGImage() {
  // Skip these problematic images (promotional/ebook images, not scenic photos)
  const skipPatterns = [
    '77016f18a7c8',  // Ebook promotional image
    'aquamare',
    'bag-collection',
    'roch-cuisine',
  ];

  const searchDirs = ['paliton-beach', 'cambugahay-falls', 'siquijor-general'];
  const candidates = [];

  for (const dir of searchDirs) {
    const dirPath = path.join(SCRAPED_DIR, dir);
    if (!fs.existsSync(dirPath)) continue;

    const files = fs.readdirSync(dirPath).filter(f =>
      /\.(jpg|jpeg|png|webp)$/i.test(f)
    );

    for (const file of files) {
      // Skip promotional/non-scenic images
      if (skipPatterns.some(pattern => file.includes(pattern))) {
        continue;
      }

      const filePath = path.join(dirPath, file);
      try {
        const metadata = await sharp(filePath).metadata();
        const ratio = metadata.width / metadata.height;

        // Prefer landscape images close to 1200x630 ratio (1.9)
        if (ratio >= 1.5 && ratio <= 2.1 && metadata.width >= 800) {
          candidates.push({
            path: filePath,
            width: metadata.width,
            height: metadata.height,
            ratioDiff: Math.abs(ratio - 1.9),
          });
        }
      } catch (e) {
        // Skip files that can't be processed
      }
    }
  }

  if (candidates.length === 0) return null;

  // Sort by closest to ideal ratio
  candidates.sort((a, b) => a.ratioDiff - b.ratioDiff);
  return candidates[0].path;
}

/**
 * Create OpenGraph image with overlay branding
 */
async function createOGImage() {
  console.log('\nCreating OG image...');

  const OG_WIDTH = 1200;
  const OG_HEIGHT = 630;

  const baseImagePath = await findBestOGImage();

  let baseImage;

  if (baseImagePath) {
    console.log(`  Using base image: ${path.basename(baseImagePath)}`);

    // Load and resize base image to cover OG dimensions
    baseImage = await sharp(baseImagePath)
      .resize(OG_WIDTH, OG_HEIGHT, {
        fit: 'cover',
        position: 'center',
      })
      .toBuffer();
  } else {
    console.log('  No suitable image found, creating gradient background');

    // Create a jungle green gradient background
    baseImage = await sharp({
      create: {
        width: OG_WIDTH,
        height: OG_HEIGHT,
        channels: 3,
        background: { r: 26, g: 77, b: 46 }, // jungle-700
      },
    })
      .jpeg()
      .toBuffer();
  }

  // Create gradient overlay SVG using brand colors (jungle green gradient like hero-gradient)
  const overlaySvg = `
    <svg width="${OG_WIDTH}" height="${OG_HEIGHT}" xmlns="http://www.w3.org/2000/svg">
      <defs>
        <linearGradient id="grad" x1="0%" y1="0%" x2="0%" y2="100%">
          <stop offset="0%" style="stop-color:${JUNGLE_700};stop-opacity:0.15" />
          <stop offset="50%" style="stop-color:${JUNGLE_700};stop-opacity:0.6" />
          <stop offset="75%" style="stop-color:${JUNGLE_700};stop-opacity:0.85" />
          <stop offset="100%" style="stop-color:${JUNGLE_700};stop-opacity:0.95" />
        </linearGradient>
      </defs>
      <rect width="100%" height="100%" fill="url(#grad)" />
    </svg>
  `;

  // Create text overlay SVG with brand typography
  const textSvg = `
    <svg width="${OG_WIDTH}" height="${OG_HEIGHT}" xmlns="http://www.w3.org/2000/svg">
      <defs>
        <style>
          @import url('https://fonts.googleapis.com/css2?family=Montserrat:wght@700&amp;family=Inter:wght@400&amp;display=swap');
        </style>
      </defs>
      <!-- Amber accent bar -->
      <rect x="60" y="510" width="80" height="4" fill="${AMBER_400}" rx="2"/>
      <!-- Title with shadow -->
      <text x="62" y="562" font-family="Montserrat, Arial, sans-serif" font-size="56" font-weight="700" fill="rgba(0,0,0,0.3)">Siquijor.xyz</text>
      <text x="60" y="560" font-family="Montserrat, Arial, sans-serif" font-size="56" font-weight="700" fill="${CREAM}">Siquijor.xyz</text>
      <!-- Subtitle -->
      <text x="60" y="600" font-family="Inter, Arial, sans-serif" font-size="24" fill="rgba(253,252,248,0.85)">Your Guide to the Mystical Island of the Philippines</text>
    </svg>
  `;

  // Composite everything together
  const overlayBuffer = Buffer.from(overlaySvg);
  const textBuffer = Buffer.from(textSvg);

  await sharp(baseImage)
    .composite([
      { input: overlayBuffer, blend: 'over' },
      { input: textBuffer, blend: 'over' },
    ])
    .jpeg({ quality: 90 })
    .toFile(path.join(PUBLIC_DIR, 'og-image.jpg'));

  console.log('  ✓ Created og-image.jpg');

  // Also save to images directory
  await sharp(baseImage)
    .composite([
      { input: overlayBuffer, blend: 'over' },
      { input: textBuffer, blend: 'over' },
    ])
    .jpeg({ quality: 90 })
    .toFile(path.join(IMAGES_DIR, 'og-default.jpg'));

  console.log('  ✓ Created images/og-default.jpg');
}

/**
 * Create web manifest
 */
function createWebManifest() {
  console.log('\nCreating web manifest...');

  const manifest = {
    name: 'Siquijor.xyz',
    short_name: 'Siquijor',
    icons: [
      {
        src: '/android-chrome-192x192.png',
        sizes: '192x192',
        type: 'image/png',
      },
      {
        src: '/android-chrome-512x512.png',
        sizes: '512x512',
        type: 'image/png',
      },
    ],
    theme_color: JUNGLE_700,
    background_color: CREAM,
    display: 'standalone',
  };

  fs.writeFileSync(
    path.join(PUBLIC_DIR, 'site.webmanifest'),
    JSON.stringify(manifest, null, 2)
  );

  console.log('  ✓ Created site.webmanifest');
}

/**
 * Main execution
 */
async function main() {
  console.log('\n' + '='.repeat(60));
  console.log('  GENERATING META ASSETS (Sharp.js)');
  console.log('='.repeat(60));

  try {
    await createFavicon();
    await createOGImage();
    createWebManifest();

    console.log('\n' + '='.repeat(60));
    console.log('  META ASSETS GENERATED SUCCESSFULLY');
    console.log('='.repeat(60) + '\n');
  } catch (error) {
    console.error('Error generating meta assets:', error);
    process.exit(1);
  }
}

main();
